### PR TITLE
Adjusts High DPI scaling to enable differential rendering

### LINKED ISSUE #1064: Terminal output not speed enough
The claims on launch were that the speed of the terminal is superb. Looking at the code it does seem very much like it, that the fastest technologies are used and the coding standards are extremely high. However, when I tried to measure these claims, I was surprised that comparing it with Fluent Terminal (which is built atop web technologies) outperforms Windows Terminal in raw output speed.

# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18362.116]
Windows Terminal version (if applicable): commit id d24d647c
Fluent Terminal: 0.4.1.0

```

# Steps to reproduce

In my favorite Ubuntu WSL instance, I issued
```powershell
Get-ChildItem /bin/,/boot/,/etc/,/home/,/lib,/lib64/,/opt/,/sbin/,/tmp/,/usr/,/var/ -Recurse -ReadOnly -ErrorAction SilentlyContinue | Out-String | New-Variable -Name Result;
Measure-Command { Write-Host $Result }
```

# Expected behavior

I'd have guessed Windows Terminal to be the fastest when compared with Fluent Terminal (my current terminal of choice) and the old-school terminal. Was there a goal for performance? This is totally a non critical issue, I just wanted to let you guys know. (Critical issue is [AltGr](https://github.com/microsoft/terminal/issues/521) combinations not working. Until that's not sorted out, I literally cannot type in the terminal.)

# Actual behavior

### Old-school terminal
```
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 9
Milliseconds      : 107
Ticks             : 91073049
TotalDays         : 0.000105408621527778
TotalHours        : 0.00252980691666667
TotalMinutes      : 0.151788415
TotalSeconds      : 9.1073049
TotalMilliseconds : 9107.3049
```

### Fluent Terminal
```
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 4
Milliseconds      : 243
Ticks             : 42433973
TotalDays         : 4.91133946759259E-05
TotalHours        : 0.00117872147222222
TotalMinutes      : 0.0707232883333333
TotalSeconds      : 4.2433973
TotalMilliseconds : 4243.3973
```

### Windows Terminal
```
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 6
Milliseconds      : 130
Ticks             : 61304207
TotalDays         : 7.0953943287037E-05
TotalHours        : 0.00170289463888889
TotalMinutes      : 0.102173678333333
TotalSeconds      : 6.1304207
TotalMilliseconds : 6130.4207
```

### LINKED ISSUE #3515: Extremely sluggish behavior when in full screen on a 4K monitor
# Environment

```none
Windows build number:
Platform      Version               VersionString
Win32NT     10.0.18362.0        Microsoft Windows NT 10.0.18362.0
Windows Terminal version: Version: 0.6.2951.0
```

# Steps to reproduce

Make the Windows Terminal go full-screen on a 4K monitor.
# Expected behavior
Fast and responsive (like it performs in non-full screen mode).

# Actual behavior
Super slow and sluggish behavior. There is at least a 2s delay for typing and using `ls` for example will result in a slow animation for listing the content in a directory. The Terminal is pretty much unusable if I go full-screen

I have a laptop with a 1080p display and I also have an external monitor connected via mini-display port. Monitor is 4K native 60 Hz, and it works without any issues. My laptop has also a dedicated GPU (nVidia GTX850m with the latest drivers). All other apps (i.e. Spotify, Microsoft Edge, Word etc) perform just fine when in full-screen on this 4K display. However, the Windows Terminal performs super slow when I go full screen on the monitor.
Terminal performs fine when I put it in full-screen on the laptop's main display.

### LINKED ISSUE #5320: Remove implicit scaling to improve performance and crispness of High DPI text
In high DPI, we have to disable differential drawing.
This is because the text is drawn at fractional pixel heights and baselines automatically as we're scaling the entire d2d render target. The fractional drawing is also making it look a little bit off for High DPI customers.

So doing the work of removing the implicit scaling and installing our own transforms on the origins and sizes of the five things we draw should resolve this issue:
1. Clear background rectangles
2. The cursor
3. Selection rectangles
4. The background color behind text
5. Text runs

### PR DESCRIPTION
**Pull request by miniksa**
## Summary of the Pull Request
- Adjusts scaling practices in `DxEngine` (and related scaling practices in `TerminalControl`) for pixel-perfect row baselines and spacing at High DPI such that differential row-by-row rendering can be applied at High DPI.

## References
- #5185 

## PR Checklist
* [x] Closes #5320, closes #3515, closes #1064
* [x] I work here.
* [x] Manually tested.
* [x] No doc.
* [x] Am core contributor. Also discussed with some of them already via Teams.

## Detailed Description of the Pull Request / Additional comments

**WAS:**
- We were using implicit DPI scaling on the `ID2D1RenderTarget` and running all of our processing in DIPs (Device-Independent Pixels). That's all well and good for getting things bootstrapped quickly, but it leaves the actual scaling of the draw commands up to the discretion of the rendering target.
- When we don't get to explicitly choose exactly how many pixels tall/wide and our X/Y placement perfectly, the nature of floating point multiplication and division required to do the presentation can cause us to drift off slightly out of our control depending on what the final display resolution actually is.
- Differential drawing cannot work unless we can know the exact integer pixels that need to be copied/moved/preserved/replaced between frames to give to the `IDXGISwapChain1::Present1` method. If things spill into fractional pixels or the sizes of rows/columns vary as they are rounded up and down implicitly, then we cannot do the differential rendering.

**NOW:**
- When deciding on a font, the `DxEngine` will take the scale factor into account and adjust the proposed height of the requested font. Then the remainder of the existing code that adjusts the baseline and integer-ifies each character cell will run naturally from there. That code already works correctly to align the height at normal DPI and scale out the font heights and advances to take an exact integer of pixels.
- `TermControl` has to use the scale now, in some places, and stop scaling in other places. This has to do with how the target's nature used to be implicit and is now explicit. For instance, determining where the cursor click hits must be scaled now. And determining the pixel size of the display canvas must no longer be scaled.
- `DxEngine` will no longer attempt to scale the invalid regions per my attempts in #5185 because the cell size is scaled. So it should work the same as at 96 DPI.
- The block is removed from the `DxEngine` that was causing a full invalidate on every frame at High DPI.
- A TODO was removed from `TermControl` that was invalidating everything when the DPI changed because the underlying renderer will already do that.

## Validation Steps Performed
* [x] Check at 150% DPI. Print text, scroll text down and up, do selection.
* [x] Check at 100% DPI. Print text, scroll text down and up, do selection.
* [x] Span two different DPI monitors and drag between them.
* [x] Giant pile of tests in https://github.com/microsoft/terminal/pull/5345#issuecomment-614127648

### CODE REVIEW

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -2094,31 +2092,22 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
-
-        // Manually multiply by the scaling factor. The DX engine doesn't
-        // actually store the scaled font size in the fontInfo.GetSize()
-        // property when the DX engine is in Composition mode (which it is for
-        // the Terminal). At runtime, this is fine, as we'll transform
-        // everything by our scaling, so it'll work out. However, right now we
-        // need to get the exact pixel count.
-        const float fFontWidth = gsl::narrow_cast<float>(fontSize.X * scale);
-        const float fFontHeight = gsl::narrow_cast<float>(fontSize.Y * scale);
+        const auto scale = dxEngine->GetScaling();
```
Comment (miniksa) [inline src/cascadia/TerminalControl/TermControl.cpp:2096]: Move back up so there's no delta.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -2312,6 +2301,10 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
     // - the corresponding viewport terminal position for the given Point parameter
     const COORD TermControl::_GetTerminalPosition(winrt::Windows::Foundation::Point cursorPosition)
     {
+        // compensate for DPI scenarios.
+        cursorPosition.X *= SwapChainPanel().CompositionScaleX();
+        cursorPosition.Y *= SwapChainPanel().CompositionScaleY();
+
```
Comment (miniksa) [inline src/cascadia/TerminalControl/TermControl.cpp:2307]: Candidate to use `til::point`?

File: src/renderer/dx/DxRenderer.cpp [hunk 3]
```diff
@@ -837,7 +834,7 @@ CATCH_RETURN();
     }
     case SwapChainMode::ForComposition:
     {
-        return _sizeTarget.scale(til::math::ceiling, _scale);
+        return _sizeTarget;
```
Comment (miniksa) [inline src/renderer/dx/DxRenderer.cpp:837]: Double-check this.

Comment (miniksa) [reply src/renderer/dx/DxRenderer.cpp:837]: Yep, removing scaling is correct. And I found a bug in resize while I was at it.

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 2 above)
Comment (miniksa) [reply src/cascadia/TerminalControl/TermControl.cpp:2307]: No because it would convert to integers too early.

**Comment by miniksa** [comment]
Dragging between different DPI monitors works very poorly. Need to look into that.

**Comment by miniksa** [comment]
> Dragging between different DPI monitors works very poorly. Need to look into that.

Latest push is better but we get invalidated multiple times so there's some flickering and missing data (partial redraw?)

**Comment by DHowett-MSFT** [comment]
Selfhost notes:

* Scaling the font with mouse wheel/pinch drops the terminal size as though it only had 1/scale% the pixels available. Zooming my terminal when my machine is 200% scale does this:

image

* I don't know if this is a regression from differential drawing, but now when I launch vim it appears and then gets blanked

image

If i make a selection, it reappears where i selected:

image

**Comment by DHowett-MSFT** [comment]
* I don't personally mind the 1px cursor and underlines

image

* Reloading the settings has the same effect on font size as zooming

**Comment by DHowett-MSFT** [comment]
* because xaml pointer coordinates are in DIPs, we need to check the distance/scaling math in PointerMoved, because it uses "25% of the cell size", but it's measuring 25% of a scaled-up cell size versus the distance in DIPs.

**Comment by DHowett-MSFT** [comment]
TSFInputControl positioning (for cursor) and sizing (for font) need to be scaled.

image

**Comment by DHowett-MSFT** [comment]
* you’ll also need to audit the wpf control for the same problems.

* the thing where vim is invisible until a selection is made is true on master too!

* the gutter still shows up uncleared if you scale the font with the wheel to a size that makes fewer lines fit on the screen

**Comment by DHowett-MSFT** [comment]
Is there some way to hide the fake font scaling inside the renderer entirely? It would make it so the consumer can just pretend it’s in DIP..

**Comment by miniksa** [comment]
@DHowett-MSFT, your overwhelming list of things makes me want to cry.

**Comment by miniksa** [comment]
- [ ] - Dragging between monitors doesn't fully refresh, receives multiple events
- [ ] - 1px cursor types will be very thin as they're not being scaled
- [ ] - Scaling font with Ctrl+mouse wheel doesn't work right
- [ ] - Vim is blanked. This is on master too.
- [ ] - Settings reload doesn't work right for font scaling
- [ ] - `PointerMoved` is having scaling issues
- [ ] - `TSFInputControl` needs to account for scaling
- [ ] - WPF control needs auditing
- [ ] - Gutter isn't cleared scaling font with wheel to size that makes fewer lines fit
- [ ] - Can we completely hide all pixel stuff into the renderer for the fake font scaling

**Comment by miniksa** [comment]
> Is there some way to hide the fake font scaling inside the renderer entirely? It would make it so the consumer can just pretend it’s in DIP..

Not without completely inverting the responsibility of hit testing into the inside of the renderer, which is a thing I've been wanting to do for a long time, but is going to be well beyond the scope of this.

**Comment by miniksa** [comment]
> > Is there some way to hide the fake font scaling inside the renderer entirely? It would make it so the consumer can just pretend it’s in DIP..
> 
> Not without completely inverting the responsibility of hit testing into the inside of the renderer, which is a thing I've been wanting to do for a long time, but is going to be well beyond the scope of this.

I tried to do something like this and it made some math errors compound. Things are casting floats into integers throughout our codebase and having an extra margin of error here by trying to give the DIPs in a float is cascading poorly.

**Comment by zadjii-msft** [comment]
Alright, I think I've got the `TSFInputControl` patched up. I'll need to add some comments but I'll make a pull into this branch

image
image
image
Yea these are jibberish, but the suggestion seems to appear at the correct location, as well as whatever UI the IME creates.

image
image

**Comment by DHowett-MSFT** [comment]
you can rebase your master merge out of existence, unless it's required for conflict resolution

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 4]
```diff
@@ -1673,22 +1675,84 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+    //        control wasn't focused as the window's DPI changed, so it only got
+    //        these messages after XAML updated it's scaling.
+    //   - 3. Finally, a CompositionScaleChanged with the _new_ DPI.
+    //   - 4. We'll usually get another SizeChanged some time after this last
+    //     ScaleChanged. This usually seems to happen after something triggers
+    //     the UI to re-layout, like hovering over the scrollbar. This event
+    //     doesn't reliably happen immediately after a scale change, so we can't
+    //     depend on it (despite the fact that both the scale and size state is
+    //     definitely correct in it)
+    // - In the 3rd event, we're going to update our font size for te new DPI.
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalControl/TermControl.cpp:1720]:
```suggestion
    // - In the 3rd event, we're going to update our font size for the new DPI.

```

**Comment by DHowett-MSFT** [comment]
> 1px cursor types will be very thin as they're not being scaled (migrie: wait is this bad? this looks crisp and beautiful IMO, but sure let's scale it)

let's keep them, they're beautiful at 1dpx

**Comment by miniksa** [comment]
> > 1px cursor types will be very thin as they're not being scaled (migrie: wait is this bad? this looks crisp and beautiful IMO, but sure let's scale it)
> 
> let's keep them, they're beautiful at 1dpx

Yeah you two have the opinion here so I'm going with it. I'll leave it.

**Comment by DHowett-MSFT** [comment]
This is glorious and robust and I love it.

**Pull request merged**